crates/quantum_info/src/pauli_lindblad_map/pauli_lindblad_map_class.rs

```diff
@@ -0,0 +1,1177 @@
+    ///
+    /// Examples:
+    ///
+    ///     .. code-block:: python
+    ///
+    ///         >>> pauli_lindblad_map = PauliLindbladMap.from_list([("IXZXYYZZ", 2.5), ("ZXIXYYZZ", 0.5)])
+    ///         >>> assert pauli_lindblad_map == pauli_lindblad_map.copy()
+    ///         >>> assert pauli_lindblad_map is not pauli_lindblad_map.copy()
+    fn copy(&self) -> PyResult<Self> {
+        let inner = self.inner.read().map_err(|_| InnerReadError)?;
+        Ok(inner.clone().into())
+    }
+
+    /// The number of qubits the map acts on.
+    ///
+    /// This is not inferable from any other shape or values, since identities are not stored
+    /// explicitly.
+    #[getter]
+    #[inline]
+    pub fn num_qubits(&self) -> PyResult<u32> {
+        let inner = self.inner.read().map_err(|_| InnerReadError)?;
+        Ok(inner.num_qubits())
+    }
+
+    /// The number of generator terms in the exponent for this map.
+    #[getter]
+    #[inline]
+    pub fn num_terms(&self) -> PyResult<usize> {
+        let inner = self.inner.read().map_err(|_| InnerReadError)?;
+        Ok(inner.num_terms())
+    }
+
+    /// The gamma for the map.
+    #[getter]
+    #[inline]
+    fn get_gamma(&self) -> PyResult<f64> {
+        let inner = self.inner.read().map_err(|_| InnerReadError)?;
+        Ok(inner.gamma)
+    }
+
+    /// The rates for the map.
+    #[getter]
+    fn get_rates(slf_: Bound<Self>) -> Bound<PyArray1<f64>> {
+        let borrowed = &slf_.borrow();
+        let inner = borrowed.inner.read().unwrap();
+        let rates = inner.rates();
+        let arr = ::ndarray::aview1(rates);
+        // SAFETY: in order to call this function, the lifetime of `self` must be managed by Python.
+        // We tie the lifetime of the array to `slf_`, and there are no public ways to modify the
+        // `Box<[u32]>` allocation (including dropping or reallocating it) other than the entire
+        // object getting dropped, which Python will keep safe.
+        let out = unsafe { PyArray1::borrow_from_array(&arr, slf_.into_any()) };
+        out.readwrite().make_nonwriteable();
+        out
+    }
+
+    /// The probabilities for the map.
+    #[getter]
+    fn get_probabilities(slf_: Bound<Self>) -> Bound<PyArray1<f64>> {
+        let borrowed = &slf_.borrow();
+        let inner = borrowed.inner.read().unwrap();
+        let probabilities = inner.probabilities();
+        let arr = ::ndarray::aview1(probabilities);
+        // SAFETY: in order to call this function, the lifetime of `self` must be managed by Python.
+        // We tie the lifetime of the array to `slf_`, and there are no public ways to modify the
+        // `Box<[u32]>` allocation (including dropping or reallocating it) other than the entire
+        // object getting dropped, which Python will keep safe.
+        let out = unsafe { PyArray1::borrow_from_array(&arr, slf_.into_any()) };
+        out.readwrite().make_nonwriteable();
+        out
+    }
```

Comment:
The safety guarantees on these really don't hold in the same way that they did in the smaller scale objects.  I'm a bit worried about this.

Reply:
Thoughts on how to resolve this?